Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+PathComponent.swift

```diff
@@ -37,88 +37,444 @@ extension PathHierarchy {
+                let kindWithoutLanguage = kind.dropFirst(languagePrefix.count)
                 return PathComponent(full: full, name: name, kind: kindWithoutLanguage, hash: hash)
             }
         }
         return PathComponent(full: full, name: name, kind: nil, hash: hash)
     }
     
-    static func joined<PathComponents>(_ pathComponents: PathComponents) -> String where PathComponents: Sequence, PathComponents.Element == PathComponent {
-        return pathComponents.map(\.full).joined(separator: "/")
+    static func split(_ path: String) -> (componentSubstrings: [Substring], isAbsolute: Bool) {
+        var components: [Substring] = self.split(path)
+        
+        let isAbsolute: Bool
+        if path.first == PathComponentScanner.separator {
+            guard let maybeModuleName = components.first?.dropFirst(), !maybeModuleName.isEmpty else {
+                return ([], true)
+            }
+            isAbsolute = maybeModuleName.isValidModuleName()
+            assert(NodeURLGenerator.Path.documentationFolderName.isValidModuleName())
+            assert(NodeURLGenerator.Path.tutorialsFolderName.isValidModuleName())
+            if isAbsolute {
+                _ = components[components.startIndex].removeFirst()
+            }
+        } else {
+            let name = components.first.map(String.init)
+            isAbsolute = name == NodeURLGenerator.Path.documentationFolderName
+                      || name == NodeURLGenerator.Path.tutorialsFolderName
+        }
+        
+        return (components, isAbsolute)
+    }
+    
+    private static func split(_ path: String) -> [Substring] {
+        var result = [Substring]()
+        var scanner = PathComponentScanner(path[...])
+        
+        let anchorResult = scanner.scanAnchorComponent()
+        
+        while !scanner.isEmpty {
+            let component = scanner.scanPathComponent()
+            if !component.isEmpty {
+                result.append(component)
+            }
+        }
+        
+        if let anchorResult = anchorResult{
+            result.append(anchorResult)
+        }
+
+        return result
+    }
+}
+
+private struct PathComponentScanner {
+    private var remaining: Substring
+    
+    static let separator: Character = "/"
+    private static let anchorSeparator: Character = "#"
+    private static let operatorEnd: Character = ")"
+    
+    init(_ original: Substring) {
+        remaining = original
+    }
+    
+    var isEmpty: Bool {
+        remaining.isEmpty
+    }
+    
+    mutating func scanPathComponent() -> Substring {
+        // If the next component is an operator, parse the full operator before splitting on "/" ("/" may appear in the operator name)
+        if remaining.unicodeScalars.prefix(3).allSatisfy(\.isValidOperatorHead) {
+            guard let operatorEndIndex = remaining.firstIndex(of: Self.operatorEnd) else {
+                defer { remaining.removeAll() }
```

Comment:
This is very clever how you call different `defer` closures depending on the contents of the string and the logic here and how you update the `remaining` value _after_ determining the return value (if I'm reading this correctly). Nice!